Adding golang govulncheck action

robertsirc opened the pull request:
I added the golang govulncheck action.

**What this PR does / why we need it**:

**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains user facing changes (the `docs needed` label should be applied if so)
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

robertsirc committed:
adding govulncheck

Signed-off-by: Robert Sirchia <rsirchia@outlook.com>

robertsirc committed:
adding workflow_dispatch to test

Signed-off-by: Robert Sirchia <rsirchia@outlook.com>

robertsirc committed:
Merge branch 'helm:main' into adding-goland-govulncheck-action

robertsirc committed:
Adding in workflow_call to test GH Actions

Signed-off-by: Robert Sirchia <rsirchia@outlook.com>

robertsirc committed:
changing the triggers to test this GH actions

Signed-off-by: Robert Sirchia <rsirchia@outlook.com>

robertsirc committed:
fixing directory for go-packages

Signed-off-by: Robert Sirchia <rsirchia@outlook.com>

robertsirc committed:
updating go version for govulncheck

Signed-off-by: Robert Sirchia <rsirchia@outlook.com>

robertsirc committed:
removing specific go version for govulncheck

Signed-off-by: Robert Sirchia <rsirchia@outlook.com>

robertsirc committed:
moving govulncheck to a seperate job

Signed-off-by: Robert Sirchia <rsirchia@outlook.com>

robertsirc committed:
restoring the original triggers that were removed for testing

Signed-off-by: Robert Sirchia <rsirchia@outlook.com>

robertsirc committed:
adding a new line at the end of the file as per the request of the maintainers

Signed-off-by: Robert Sirchia <rsirchia@outlook.com>

gjenkins8 commented on .github/workflows/build-test.yml:
```diff
@@ -24,3 +24,15 @@ jobs:
+        with:
+          go-version: '1.22'
+      - name: govulncheck
+        uses: golang/govulncheck-action@dd0578b371c987f96d1185abb54344b44352bd58 # pin@1.0.3
```
To put it on the record, I think there is a limitation in the way this action is implemented. That is, the plugin will alert on any CVE present in a change/PR. Whether the change introduced the CVE or not. It will also _not_ alert if a CVE is discovered.

So what I think is actually desirable:
1. Changes (ie. PRs) should blocked, if and only if they introduce a vulnerability. And not because they carry the existing dependency set with a pre-existing vulnerability. This in effect requires subtracting existing vulnerabilities from the target branch from the change branch.
2. If a CVE/vulnerability is discovered. There should be a proactive alert and remediation, without waiting for an unrelated PR. Dependabot serves most of this purpose today. But to my knowledge it doesn't integrate with govulncheck to e.g. perhaps more aggressively promote PRs which contain actual fixes (it uses the standard dependency graph mechanism to find updates). And to this point, it waits until there is an upstream package fix before it does anything. Where perhaps there could e.g. be a workaround.

So, even with the above, we might still want to utilize govulncheck here. Since it might be better than nothing. We (maintainers) probably want to be able to override/ignore any findings it reports (since they may be unrelated to the change on hand)

(and to note, this is based all on my understanding of the way the govulncheck action works. Which could be wrong of course)

I could/should also at least raise an issue on the govulncheck action. To see if some of these could be idea couldbe implemented (if I'm not mistaken about the workings here)

scottrigby commented:
Also see https://github.com/helm/helm/pull/12045

robertsirc committed:
adding new workflows for govulncheck

Signed-off-by: Robert Sirchia <rsirchia@outlook.com>

robertsirc committed:
removing govulncheck from build-test

Signed-off-by: Robert Sirchia <rsirchia@outlook.com>

robertsirc committed:
changing trigger file from go.sum to go.mod

Signed-off-by: Robert Sirchia <rsirchia@outlook.com>

robertsirc committed:
adding a line break to test the GH trigger

Signed-off-by: Robert Sirchia <rsirchia@outlook.com>

robertsirc committed:
removing line break used for testing

Signed-off-by: Robert Sirchia <rsirchia@outlook.com>

robertsirc committed:
changing the trigger file

Signed-off-by: Robert Sirchia <rsirchia@outlook.com>

robertsirc commented:
@gjenkins8 and @mattfarina I updated this corrected this PR. I removed govulncheck from buid-test, added two more action. First a daily action that runs once a day to check for vulnerabilities. The second will check for vulnerabilities only if `go.sum` has a change to it.

robertsirc committed:
adding new lines at the end of each files

Signed-off-by: Robert Sirchia <rsirchia@outlook.com>

robertsirc committed:
refectoring to ONE GH action

Signed-off-by: Robert Sirchia <rsirchia@outlook.com>

mattfarina reviewed (approved):
Note,

gjenkins8 reviewed (approved):
Nice, thanks for updating!
